modules/@apostrophecms/log/index.js

```diff
@@ -0,0 +1,403 @@
+//       '*': {
+//         severity: [ 'warn', 'error' ]
+//         // match all severity levels
+//         // severity: '*'
+//         // match event types
+//         // events: [ 'event-type-1', 'event-type-2' ]
+//         // match all event types
+//         // events: '*'
+//       },
+//       // Log specific event types from the login module
+//       '@apostrophecms/login': {
+//         events: [ 'incorrect-username', 'incorrect-password' ]
+//         // match all event types
+//         // events: '*'
+//         // match specific severity levels
+//         // severity: [ 'info' ]
+//         // match all severity levels
+//         // severity: '*'
+//       }
+//     }
+//   }
+// }
+// ```
+// In this example, all errors and warnings from any module, but
+// only the specific event types (no matter the severity) from the login
+// module, are logged. The logs will be kept if *either* criterion is met.
+// `filter['*'] = true` enables logging of all events from all modules.
+//
+// ## Environment Variables
+//
+// ### `APOS_FILTER_LOGS`
+//
+// If set, this environment variable overrides the `filter` option.
+// Example:
+// ```sh
+// # same as the `filter` example above
+// export APOS_FILTER_LOGS='*:severity:warn,error;@apostrophecms/login:events:incorrect-username,incorrect-password'
+// # log everything, analogous to `{ filter: { '*': true }}`
+// export APOS_FILTER_LOGS='*'
+// ```
+//
+
+const _ = require('lodash');
+
+module.exports = {
+  options: {
+    alias: 'structuredLog'
+  },
+  init(self) {
+    self.filters = {};
+    self.initFilters();
+  },
+  methods(self) {
+    return {
+      // Normalize the filters. Detect configuration and set defaults
+      // per the current NODE_ENV if needed.
+      // Convert `{ *: true }` to an object with all severity levels.
+      // Convert severity wildcards to arrays of (all) severity levels. This
+      // speeds up the severity detection (no wildcards match).
+      // Convert eventType wildcards to `[ '*' ]` array.
+      // Override the configuration with the `APOS_FILTER_LOGS` environment variable
+      // if set.
+      initFilters() {
+        self.filters = self.options.filter || {};
+        if (process.env.APOS_FILTER_LOGS) {
+          try {
+            self.filters = self.parseEnvFilter(process.env.APOS_FILTER_LOGS);
+          } catch (e) {
+            throw new Error(`Invalid APOS_FILTER_LOGS environment variable: ${e.message}`);
+          }
+        }
+        self.filters['*'] = self.filters['*'] || {};
+
+        // Transform *: true - log absolutely everything.
+        if (self.filters['*'] === true) {
+          self.filters['*'] = {
+            severity: self.getDefaultSeverity(false)
+          };
+          return;
+        }
+        // Add environment specific severity levels if no severity is specified.
+        if (!self.filters['*'].severity || self.filters['*'].severity.length === 0) {
```

Comment:
Empty array should not mean "everything." It's legitimate to create configuration via tools and tools are likely to emit empty arrays when they mean "nothing" rather than skipping a property.